Rework blockchain import

holiman opened the pull request:
This PR addresses some shortcomings in the blockchain downloader and block processing, specifically the behaviour around (invalid) long sidechains, as happened during the Ropsten constantinople HF.

holiman commented:
This has now been running both full and fast sync on the benchmarking machines, and appear to be able to complete a full sync. 
The graph below shows the fast-sync network IO _after_ fast sync completion. Whereas ingress is almost the same, the ingress traffic is reduced heavily, almost to 50%. I _believe_ that is because when we sync with a new peer, we now correclty determine the common ancestor at block `N` instead of block `N-15`, and do not have to fetch 15 full blocks unnecessarily.   
![network_io](https://user-images.githubusercontent.com/142290/47683330-8e7bee80-dbcf-11e8-9d01-189880cc2af2.png)

The graphs also shows that block processing is faster in this PR -- but that's not really fair, because this PR handles some errors before 'starting the clock' -- and we only really measure the time _execution_ of a block takes.

karalabe commented on eth/downloader/downloader.go:
```diff
@@ -587,41 +587,87 @@ func (d *Downloader) fetchHeight(p *peerConnection) (*types.Header, error) {
+	if requestBottom < 0 {
+		requestBottom = 0
+	}
+	totalSpan := requestHead - requestBottom
```
What happens here if our local height is larger than the remote one? Total Span will be negative. Is that acceptable?

holiman replied:
That works, although I agree it's not obvious. But it's in the testcases:
```
		{1000, 1500,
			[]int{997, 999},
```
So remote `1000`, local `1500`, will request `997, 998`.

karalabe commented on eth/downloader/downloader.go:
```diff
@@ -587,41 +587,87 @@ func (d *Downloader) fetchHeight(p *peerConnection) (*types.Header, error) {
 	}
 }
 
+// calcRequestSpan calculates what headers to request from a peer when trying to determine the
```
The name of the method in the doc doesn't match it's actual name. Please also document what the return parameters are. They are 4 ints, which is very non obvious as to what means what.

karalabe commented on eth/downloader/downloader_test.go:
```diff
@@ -1753,3 +1754,81 @@ func testDeliverHeadersHang(t *testing.T, protocol int, mode SyncMode) {
+		{15000, 13006,
+			[]int{14823, 14839, 14855, 14871, 14887, 14903, 14919, 14935, 14951, 14967, 14983, 14999},
+		},
+		//Remote is pretty close to us. We don't have to fetch as many
```
Please add a space after the comment slashes.

karalabe commented on eth/downloader/downloader_test.go:
```diff
@@ -1753,3 +1754,81 @@ func testDeliverHeadersHang(t *testing.T, protocol int, mode SyncMode) {
+		{1500, 1500,
+			[]int{1497, 1499},
+		},
+		// We're higher than the remote! Odd
```
It can happen during Clique where we have 2x 1-diff blocks and a remote side has 1x 2-diff block.

karalabe commented on eth/downloader/downloader.go:
```diff
@@ -587,41 +587,87 @@ func (d *Downloader) fetchHeight(p *peerConnection) (*types.Header, error) {
+	totalSpan := requestHead - requestBottom
+	span := 1 + totalSpan/MaxCount
+	if span < 2 {
+		span = 2
```
An particular reason for putting the minimum span to 2? Maybe if we're close enough we can request head-1 and head-2 instea of head-3? Not sure, thinking out aloud here really.

holiman replied:
`span` is not the same as `skip`. So `span` of `2` means we're getting every `2`nd item, so `skip` is `1`.

karalabe commented on eth/downloader/downloader.go:
```diff
@@ -647,30 +693,31 @@ func (d *Downloader) findAncestor(p *peerConnection, height uint64) (uint64, err
+					//if number > remoteHeight && i == limit-1 {
+					//	p.log.Warn("Lied about chain head", "reported", requestHead, "found", number)
+					//	return 0, errStallingPeer
+					//}
```
Might as well delete this code block.

karalabe commented on eth/downloader/downloader_test.go:
```diff
@@ -1753,3 +1754,81 @@ func testDeliverHeadersHang(t *testing.T, protocol int, mode SyncMode) {
 }
+
+func TestRemoteHeaderRequestSpan(t *testing.T) {
+
```
Please delete this empty line.

karalabe commented on eth/downloader/downloader_test.go:
```diff
@@ -1753,3 +1754,81 @@ func testDeliverHeadersHang(t *testing.T, protocol int, mode SyncMode) {
+		},
+		{15000, 13006,
+			[]int{14823, 14839, 14855, 14871, 14887, 14903, 14919, 14935, 14951, 14967, 14983, 14999},
+		},
```
Seems weird that in normal mode of operation we would request headers we most certainly don't have yet. Does it really make sense to ask for ancestors way out of our chain height?

holiman replied:
That's a very good question. In non-normal mode of operation, there's this odd thing that may happen:
```
      common ancestor A (1000)
                    /                     \
    ourchain B(1001)      otherchainC(1001)
       |                                      |
       |                                      | 
    ourhead E(1200)        otherblock G(1200)
                                 otherhead G(1500)
```

So with the current downloader logic, the first time it sees the (bad) sidechain, it will 

* download from `A` to `G`, and try to insert 500 blocks ( or so) . 
* If ancestors are pruned, it will write `C` to `G` to disk, and then fail when it tries to process `G` for `td`-reasons. 

Next time, it will
* mistakenly believe `G` (or a few blocks before) is the common ancestor, since it's present on disk. It will again download `G` to `F`, and fail again. 
It will then repeat this pattern over and over again. 

The change in the downloader is to _not_ view common ancestor as indeed an ancestor, but merely 'the latest block that we already have on disk'. Coupled with the changes in block processor, the new behaviour at the scenario above is. 
 
* download from `A` to `G`, and try to insert 500 blocks ( or so) . 
* If ancestors are pruned, it will write `C` to `F` to disk, and then fail when it tries to process `G` for `td`-reasons. 

Next time, it will
* identify `F` (or some block right before) as the common ancestor, and only download a couple of blocks. 

(It will then repeat this pattern over and over again still)


So how will it behave in common scenarios? 
If we're synching from scratch, it will fail, and have to fall back to a binary search. However, the binary search is actually not that bad, there are more roundtrips but at least we hit the exact right block, instead of missing and downloading ~10 extra blocks.

holiman replied:
Separating the downloader-change out from the blockchain import change might require changing this behaviour, however, so that the span of blocks always contain our own height.

karalabe commented:
@holiman Could you please rebase this on master? Thx

holiman commented:
done

karalabe commented on eth/downloader/downloader_test.go:
```diff
@@ -1567,3 +1567,80 @@ func TestRemoteHeaderRequestSpan(t *testing.T) {
 	}
 }
+
+func TestRemoteHeaderRequestSpan(t *testing.T) {
```
This was already merged in, wasn't it? Why does it still appear in this PR?

karalabe commented on core/blockchain.go:
```diff
@@ -1082,164 +1169,244 @@ func (bc *BlockChain) insertChain(chain types.Blocks) (int, []interface{}, []*ty
+	batch := newBatch(chain, results, bc.Validator())
+
+	if done, blockError := batch.NextErr(); blockError != nil {
+		if blockError == consensus.ErrPrunedAncestor {
```
Please separate out everything at this nesting level into an `insertSidechain` method. This all got humonguous already and very deeply nested. Should help clean things up a lot.

holiman commented on eth/api.go:
```diff
@@ -248,6 +249,44 @@ func (api *PrivateAdminAPI) ImportChain(file string) (bool, error) {
 
+// SlowImportChain imports a blocks from a local file, with one block every 15 seconds
+// this is a debug helper that can be used to simulate blocks arriving from some external source
+func (api *PrivateDebugAPI) SlowImportChain(file string) (bool, error) {
```
No I agree

holiman commented on core/chain_makers_test.go:
```diff
@@ -28,6 +35,84 @@ import (
+		}
+	}
+	return nil
+}
```
I can remove it, no worries

holiman commented:
I think I have addressed all concerns now. However, after the latest changes, the code is relatively untested. Once you are satisfied with the general 'look' of the code, we should either add testcases that cover long (bad-state) sidechains, or run some external tests on that scenario.

karalabe commented on core/blockchain.go:
```diff
@@ -1107,139 +1239,184 @@ func (bc *BlockChain) insertChain(chain types.Blocks) (int, []interface{}, []*ty
+		}
+		parents = append(parents, parent)
+		parent = bc.GetBlock(parent.ParentHash(), parent.NumberU64()-1)
+	}
```
Copy paste from discord:

---

I found a really weird behavior in the current chain importer
my node crashed, so it lost the state of many blocks
I restarted it
and was wondering why the memory keeps growing
now the bummed out part
the downloader fetches the first 2 blocks, calls InsertChain on them
and the chain inserter kee[s importing all the blocks it previously had, reprocessing everything
even though it only got 2 blocks as a parameter
apparently some code detects as if we just made a reorg onto a longer side chain
and start reprocessing all the blocks
which might explain some oom issues on restart/rinkeby
I've no idea though how the chain inserert gets into this fucked up phase
ooooh, it's the new downloade rlogic
even in full sync mode, it pulls the head, and detects the "sidechian" that I previously had
and when a side chain wins, we load all the blocks into memory currently
whch explains the blowup in the memory consumption

```
            // Competitor chain beat canonical, gather all blocks from the common ancestor
            var winner []*types.Block

            parent := bc.GetBlock(block.ParentHash(), block.NumberU64()-1)
            for !bc.HasState(parent.Root()) {
                winner = append(winner, parent)
                parent = bc.GetBlock(parent.ParentHash(), parent.NumberU64()-1)
            }
            for j := 0; j < len(winner)/2; j++ {
                winner[j], winner[len(winner)-1-j] = winner[len(winner)-1-j], winner[j]
            }
            // Import all the pruned blocks to make the state available
            bc.chainmu.Unlock()
            _, evs, logs, err := bc.insertChain(winner)
            bc.chainmu.Lock()
            events, coalescedLogs = evs, logs

            if err != nil {
                return i, events, coalescedLogs, err
            }
```

this code snipped became quite dangerous with the downloader update
we need to control the total memory consumption here
one solution would be to recomstruct the hash chain only, and then load+import in batches

karalabe commented on eth/downloader/api.go:
```diff
@@ -40,8 +40,8 @@ type PublicDownloaderAPI struct {
-		d:                         d,
-		mux:                       m,
+		d:   d,
+		mux: m,
```
Please update your Go installation. This is a formatting chang ein Go 1.11.
